### ISSUE TITLE
[Form] Make Post-Redirect-Get pattern possible

### ISSUE THREAD

**Issue by BabakMN** [symfony/symfony#22054]
Is it possible to use the post-redirect-get pattern with forms?

In other words, serialize and store the form in the session so if a form has errors you can still redirect back to the same page and still have access to the form for displaying values and error messages (instead of having to return a `200`).

| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.2.6

**Comment by Nek-** [comment]
I think the usage most people do is:
* if there is an error, no redirection and show the form with errors
* if there is no error redirection (what you may call "post-redirect-get")

If you want to redirect and save the form data in the session, you can. But as it's not a standard use case (and some people would say it's a wrong one) Symfony will probably never provide tools to do it out of the box.

Please correct me if I misunderstood your mind.

**Comment by BabakMN** [comment]
If we generally accept that post-redirect-get is better than just simply returning a response (as the Symfony documentation suggests, due to the form re-submission issue and so on).

Then I'm not sure why the behaviour should be different for the case when the form submission is successful and when it has errors?

It appears that the problems that PRG attempts to solve apply to any form submission whether it was successful or not.

I'd argue that it is even more important in the case when the submission fails.

For example if you are on `/login` and you post your login attempt data to the same page `/login`. If it is successful then it only makes sense and is easy and natural to do a redirect to let's say `/account`.

Otherwise you'd end up displaying the account page under the `/login` URL. Or you'd have to make your login form make the submission to the "destination" (`/account`) which is bad.

So we redirect in that case instead and it's easy and natural and it makes sense.

So my point is even though in the example above we are using PRG,  It's not a particularly good example of why PRG matters.

However in case of errors it's easy to just render and return the same page but then you will have the problems that PRG tries to solve in the first place, so that's where it matters to make the effort to use PRG.

Based on what I have seen it seems like adding form submission errors to the session flash and displaying them on the next `GET` request is pretty popular.

This can be done fairly easily if you are simply dealing with an array of error messages.

However it becomes difficult if you are using `Form` components because you will lose benefits such as the ability to pre-fill the form values (not lose what the user had typed).

So in conclusion, it appears to me that the reason that this feature is not available right now is not because using PRG is bad or pointless when the form has submission errors, but rather because it is difficult and inconvenient to serialize the necessary objects into the session so it's easier to only do PRG on success and ignore it in case of errors.

I'd like to ask others, if you had the choice to use PRG in both cases (success & failure) without any extra effort or difficulty would you go out of your way to make sure you only do it in case of success but not failure? If so why?

Thanks.

**Comment by xabbuh** [comment]
Resubmitting the form when it had errors usually is no problem. The user will just see the same form again with the same error messages. However, resubmitting the form when it was successfully processed means that you will have to take extra care to not create new records twice, send confirmation mails several times and so on. That's the reason why you usually return a redirect response when a form was successfully processed.

**Comment by BabakMN** [comment]
Thanks @xabbuh. Closing.
